### PR TITLE
Use set to represent kinds in KindForGroupVersionKinds

### LINKED ISSUE #76369: Use set to represent targets in KindForGroupVersionKinds
**What would you like to be added**:
Currently targets is an array of GroupVersionKind.
bestMatch iterates over the array to find match.
We can use set as representation of the targets so that finding match becomes O(1) operation.

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Currently kinds is an array of GroupVersionKind in KindForGroupVersionKinds.
bestMatch iterates over the array to find match.
We can use set as representation of the kinds so that finding match becomes O(1) operation.

**Which issue(s) this PR fixes**:
Fixes #76369

```release-note
NONE
```

### CODE REVIEW

**Comment by tedyu** [comment]
cc @liggitt @deads2k

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go [hunk 1]
```diff
@@ -249,16 +249,18 @@ type GroupVersions []GroupVersion
...
 func (gvs GroupVersions) KindForGroupVersionKinds(kinds []GroupVersionKind) (GroupVersionKind, bool) {
-	var targets []GroupVersionKind
+	var targets map[GroupVersionKind]struct{}
 	for _, gv := range gvs {
 		target, ok := gv.KindForGroupVersionKinds(kinds)
 		if !ok {
 			continue
 		}
-		targets = append(targets, target)
+		targets[target] = struct{}{}
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: Making this a map loses the preferred ordering of gvs

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go [hunk 2]
```diff
@@ -268,15 +270,16 @@ func (gvs GroupVersions) KindForGroupVersionKinds(kinds []GroupVersionKind) (Gro
...
-			if k == gvk {
-				return k
-			}
+func bestMatch(kinds []GroupVersionKind, targets map[GroupVersionKind]struct{}) GroupVersionKind {
+	for _, k := range kinds {
+		if _, ok := targets[k]; ok {
+			return k
 		}
 	}
-	return targets[0]
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: This previously honored the order of gvs. As a map, it no longer does.

**Review by liggitt — COMMENTED** [review]
If this doesn't fail a unit test, I think it means we need more unit tests around this, specifically for honoring the preferred order of the GroupVersions

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go (hunk 2 above)
Comment (tedyu) [reply staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: Thanks for the reminder - which doesn't seem to be recorded in existing comment in code.

What if we produce a map[GroupVersionKind]struct{} for kinds ?
that way the order is preserved for targets.

**Comment by tedyu** [comment]
@liggitt 
I have kept targets as array and used map for kinds to speed up matching.

Please take another look.

Thanks.

**Comment by liggitt** [comment]
do we have evidence that this is currently an issue. If I recall correctly, the number of kinds/targets we're iterating over is typically 1-3. I'm not sure a map allocation here is worth the cost.

**Comment by tedyu** [comment]
@liggitt 
What do you think of the simplification in KindForGroupVersionKinds() around line 200 ?
No additional map is involved for that change.

**Comment by yliaog** [comment]
since this PR is about improving performance, is there profiling result to show the performance improvement?

**Comment by tedyu** [comment]
Do you have recommendation on benchmark suite which should be run to obtain the performance result ?

Thanks

**Comment by smarterclayton** [comment]
We definitely avoided map allocations in this code path.  This method is deliberately kept small to minimize allocations, and a map would go against that.

**Pull request closed without merging**